include/wil/registry.h

```diff
@@ -17,16 +17,2855 @@
 
 #include <winreg.h>
 #include <new.h> // new(std::nothrow)
-#include "resource.h" // unique_hkey
+#include "registry_helpers.h"
+#include "resource.h"
+
+#ifndef __WIL_WINREG_
+#error __WIL_WINREG_ is required for ::wil::unique_hkey support; ensure the raw registry APIs (like RegCreateKeyExA) are available
+#endif
+
+// wil registry does not require the use of the STL or C++ exceptions (see _nothrow functions)
+// wil registry natively supports std::vector and std::wstring when preferring those types
+// wil registry uses the __WIL_WINREG_STL define to enable support for wil::shared_* types (defined in resource.h)
 
 namespace wil
 {
-    //! The key name includes the absolute path of the key in the registry, always starting at a
-    //! base key, for example, HKEY_LOCAL_MACHINE.
-    size_t const max_registry_key_name_length = 255;
+    namespace reg
+    {
+        /**
+         * \brief Helper function to translate registry return values if the value was not found
+         * \param hr HRESULT to test from registry APIs
+         * \return boolean if the HRESULT indicates the registry value was not found
+         */
+        constexpr bool is_registry_not_found(HRESULT hr) WI_NOEXCEPT
+        {
+            return (hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND)) ||
+                (hr == HRESULT_FROM_WIN32(ERROR_PATH_NOT_FOUND));
+        }
+
+        /**
+         * \brief Helper function to translate registry return values if the buffer was too small
+         * \param hr HRESULT to test from registry APIs
+         * \return boolean if the HRESULT indicates the buffer was too small for the value being read
+         */
+        constexpr bool is_registry_buffer_too_small(HRESULT hr) WI_NOEXCEPT
+        {
+            return hr == HRESULT_FROM_WIN32(ERROR_MORE_DATA);
+        }
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        /**
+         * \brief Opens a new HKEY to the specified path - see RegOpenKeyExW
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param path A string to append to the HKEY to attempt to open
+         *        can be nullptr if not needed
+         * \param access The requested access desired for the opened key
+         * \return A wil::unique_hkey containing the resulting opened HKEY
+         * \exception std::exception (including wil::ResultException) will be thrown on all failures
+         */
+        inline ::wil::unique_hkey open_unique_key(HKEY key, _In_opt_ PCWSTR path, ::wil::reg::key_access access = ::wil::reg::key_access::read)

```

Comment:
I'm forcing namespace scoping from the root global namespace to avoid a documented issue where callers can slice namespaces causing collisions (it's only happened to me twice, and they were not fun). so it seems best to guarantee we are calling exactly what we intend to call :)